### PR TITLE
added bucket library design to readme

### PR DESCRIPTION
**Pull request by jeffvance**
New README.md
Moved previous readme to new _hack/_ dir.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,27 +1,395 @@
...
+the _contract_ made to app developers regarding the artifacts of bucket creation (Secrets,
+ConfigMaps) are guaranteed. The actual creation of buckets belongs to each object store
+provisioner. However, as is true for dynamic Persistent Volume provisioning, each bucket
+provisioner needs to adhere to the inferfaces defined in the bucket library.
 
-Then install the dependencies
+An `ObjectBucketClaim` (OBC) is similar in usage to a Persistent Volume Claim
+and an `ObjectBucket` (OB) is the Persistent Volume equivalent. 
+Bucket binding refers to the actual bucket being created by the underlying object
+store provide, and the generation of artifacts which will be consumed by application pods.
```
Comment (copejon) [inline README.md]: s/provide/provider

File: README.md [hunk 2]
```diff
@@ -1,27 +1,395 @@
...
+An OBC is namespaced and references a storage class which defines the object store. The
+details of the object store (ceph, minio, cloud, on-prem) are not visible to the app pod
+and can change without disturbing the app in anyway. An OB is non-namespaced (global),
```
Comment (copejon) [inline README.md]: > can change without disturbing the app in anyway.

This is maybe a little unclear.  It reads to me as if the object store could be hot-swapped during the app's runtime.

File: README.md [hunk 3]
```diff
@@ -1,27 +1,395 @@
...
+provisioners will be simple-to-write operators because the bucket provisioning lib
+handles the bulk of the work. Each provisioner is only responsible for writing
+`Provision()` and `Delete()`functions (the _business logic_) and a short `main()`
+function. **Note:** even though the PV-PVC design supports static provisioning, only
+dynamic provisioning is supported by the bucket lib.
 
-## Update generated code
+The `Provision()` and `Delete()`functions are interfaces defined in the bucket library.
+All bucket provisioners are required to return an OB struct and the bucket
+credentials (access-key, endpoint) as defined in the bucket lib. Based on the returned
```
Comment (copejon) [inline README.md]: > (access-key, endpoint)

endpoint is defined in the OB

File: README.md [hunk 4]
```diff
@@ -1,27 +1,395 @@
...
+OB, the lib will create the bucket artifacts: user Secret, user ConfigMap, and OB
```
Comment (copejon) [inline README.md]: > user Secret, user ConfigMap

may be better to say "Credentials Secret and Bucket ConfigMap"

File: README.md [hunk 5]
```diff
@@ -1,27 +1,395 @@
...
+controller (primarily for status updates). The Secret and ConfigMap have deterministic
```
Comment (copejon) [inline README.md]: >  and OB controller (primarily for status updates)

This reads as though the OB controller(s?) is created in response to an OBC create.

File: README.md [hunk 6]
```diff
@@ -1,27 +1,395 @@
...
+
+**Note:** the bucket library has no mechanism to prevent an OBC from being deleted when one or
+more pods indirectly reference the OBC via the Secret and ConfigMap. This concept came late
+for PVCs, see
+[merged pr](https://github.com/kubernetes/community/pull/1174/files), and may be even more
+difficult to implement for OBCs.
+
+### Bucket Sharing
+
+Env though there is a 1:1 mapping of an OBC and OB, a bucket can still be shared,
```
Comment (jeffvance) [inline README.md]: typo! Also, no need for 1:1 mapping part of this sentence...

File: README.md [hunk 7]
```diff
@@ -1,27 +1,395 @@
...
+which represents this bucket. The OB's status will indicate that the related OBC
+has been deleted so that an admin has better visibility into buckets that are missing
+their connection information.
+
+The generated OB's `ownerReference` is set to the object store, not to the OBC.
+This allows an OBC to be deleted while preserving the OB. When the object store
+service is deleted all OBs belonging to that object store will be automatically
+deleted by Kubernetes due to the `ownerReference` setting.
+
+**Note:** it is an expected enhancement to support a Storage Class' _reclaimPolicy_,
```
Comment (jeffvance) [inline README.md]: delete this note. no need to try to predict the future

File: README.md [hunk 8]
```diff
@@ -1,27 +1,395 @@
...
+
+## API Specifications
+
+### OBC Custom Resource Definition
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: objectbucketclaims.store-operator.s3
```
Comment (copejon) [inline README.md]: s/store-operator.s3/objectbucket.io

File: README.md [hunk 9]
```diff
@@ -1,27 +1,395 @@
...
+the Object Store service URL.
+1. tenant allows users to define a tenant in an object store in order to namespace their buckets
+and access keys.
+
+### OBC Custom Resource (Status Updated)
+
+ ```yaml
+apiVersion: store-operator.s3/v1alpha1
+kind: ObjectBucketClaim
+...
```
Comment (copejon) [inline README.md]: add `storageClassName` as outer scope field.

File: README.md [hunk 10]
```diff
@@ -1,27 +1,395 @@
...
+    listKind: ObjectBucketList
+    plural: objectbuckets
+    singular: objectbucket
+  scope: Namespaced
+  version: v1alpha1
+  subresources:
+    status: {}
+```
+
+ ### Generated OB Custom Resource
```
Comment (copejon) [inline README.md]: delete leading whitespace

File: README.md [hunk 11]
```diff
@@ -1,27 +1,395 @@
...
+
+### Quota
+
+S3 bucket size cannot be specified; however, bucket size can be monitored in S3. The number
+of buckets can be controlled by a resource quota once
+[this k8s pr](https://github.com/kubernetes/kubernetes/pull/72384) is merged. Until then, 
+Resource Quotas cannot yet be defined for CRDs and, thus, there is no quota on the 
+number of buckets.
+
+## API Specifications
```
Comment (jeffvance) [inline README.md]: add subject to change since very new atm...

**Pull request merged**